MNTOR-1889: Add “Find exposures” step to onboarding and wire up backend

flozia opened the pull request:
<!-- The following is intended to be helpful to you. Feel free to remove anything that is not. -->

# References: 
Jira: [MNTOR-1889](https://mozilla-hub.atlassian.net/browse/MNTOR-1889)
Figma: https://github.com/mozilla/blurts-server/compare/MNTOR-1889-Wireup-enter-info-backend?expand=1


<!-- When adding a new feature: -->

# Description

This PR adds the “Find exposure” step to the onboarding flow.

# How to test

1. If you already [created `locationAutocompleteData.json` locally](https://github.com/mozilla/blurts-server#prerequisites) you can skip this step:
```
npm run create-location-data
```

2. Add the new OneRep variables in your `.env`:
```
ONEREP_API_KEY=[add-sandbox-key-here]
ONEREP_API_BASE=https://api.onerep.com
# Can be empty for testing this PR
ONEREP_WEBHOOK_SECRET=
ONEREP_DATA_BROKER_COUNT=190
```

4. If you already have run a scan you can delete the entries from your local database:
```
delete from onerep_scans; delete from onerep_profiles; update subscribers set onerep_profile_id = null;
```

5. In case you have not run the migration script locally for a while you can do so with:
```
npm run db:migrate
```

6. Visit `/redesign/user/welcome` and go through the welcome onboarding.

flozia commented on src/app/api/v1/user/welcome-scan/create/route.ts:
```diff
@@ -0,0 +1,89 @@
+        await setOnerepProfileId(subscriber, profileId);
+        await setProfileDetails(profileId, profileData);
+
+        // Start exposure scan
+        const scan = await createScan(profileId);
+        const scanId = scan.id;
+        await setOnerepScan(profileId, scanId);
```
@rhelmer Is it fine to start the scan directly after submitting the profile here?

rhelmer replied:
I just put it in the `scanning` route in my demonstration so the progress meter would stand for *something*, but it could happen here as well.

flozia commented on src/app/api/v1/user/welcome-scan/create/route.ts:
```diff
@@ -0,0 +1,89 @@
+  if (typeof session?.user?.email === "string") {
+    try {
+      const subscriber = await getSubscriberByEmail(session.user.email);
+      // FIXME check if scan has already been performed and return early if so.
```
@rhelmer Is this still something we need to fix?

flozia replied:
So it seems like we already sufficiently check this, so I removed the comment in [31a6f65](https://github.com/mozilla/blurts-server/pull/3222/commits/31a6f650df5673b5598881022f875d0ea625d83b).

flozia commented on src/app/(proper_react)/redesign/(authenticated)/user/welcome/FindExposures.tsx:
```diff
@@ -0,0 +1,72 @@
+
+  const progressSteps = 6;
+  const maxProgress = 100;
+  const totalBreachesCount = 672;
```
@rhelmer @mansaj Do you know how we can get the total breaches?

rhelmer replied:
`getAllBreaches` from `./src/db/tables/breaches.js` could be used for this, @mansaj is there a better way? I wonder if exposing a function to just return the count would be beneficial.

mansaj replied:
`getAllBreaches` would be an overkill for this (also since this is called every time?!).. we should create a util func in that file to return the count if necessary

If we aren't caching, we should really be caching the number

flozia replied:
Sorry, the screen I got this from had an error: We actually need to show the total number of data brokers here. Can we get that via API or should we set that in an `.env` for now? Same here: https://github.com/mozilla/blurts-server/blob/main/src/app/(proper_react)/(onboarding)/user/welcome/GetStarted.tsx#L75.

flozia replied:
Ok, just checked with PM and we’ll show both. Adding that as well.

flozia commented on src/app/(proper_react)/redesign/(authenticated)/user/welcome/FindExposures.tsx:
```diff
@@ -0,0 +1,72 @@
+      setScanProgress(Math.min(nextProgress, maxProgress));
+
+      // Check the scan progress and if it’s finished.
+      if (!checkingScanProgress && !scanFinished) {
+        setCheckingScanProgress(true);
+        void fetch("/api/v1/user/welcome-scan/progress")
+          .then((response) => response.json())
+          .then((result) => {
+            if (result.status && result.status === "finished") {
+              setScanFinished(true);
+            }
+            setCheckingScanProgress(false);
+          })
+          .catch((_) => setCheckingScanProgress(false));
+      }
```
@rhelmer Should we be only calling this route locally for dev?

rhelmer replied:
I think we still need to show the scan progress in either case, it's this part specifically where `getScanDetails` is called and then the results stored that we don't need to do in non-dev: https://github.com/mozilla/blurts-server/pull/3219/files#diff-2bb7d4c5a17c8bc653a51e2717446d7aeb2dc9be92711b8dfa642189b602ae13R48

flozia commented on src/app/api/v1/user/welcome-scan/progress/route.ts:
```diff
@@ -0,0 +1,76 @@
+
+        // For development we are periodically checking the scan progress and
+        // set the result if finished.
+        if (
+          scan.status === "finished" &&
+          process.env.NODE_ENV === "development"
+        ) {
+          const results = await listScanResults(profileId, { per_page: 100 });
+          await setOnerepScanResults(profileId, scan.id, {
+            data: results.data,
+          });
+        }
```
@rhelmer I expect we’ll need to add a new route for setting the scan results that can be called from the webhook?

flozia commented on src/app/(proper_react)/redesign/(authenticated)/user/welcome/EnterInfo.tsx:
```diff
@@ -102,17 +146,41 @@ export const EnterInfo = (props: Props) => {
-        "onboarding-enter-details-input-error-message"
-      ),
+      // TODO: Localize string
+      errorMessage: "You have to be at least 13 years old",
```
Note: Need to sync with Content.

pdehaan replied:
Interesting... Doesn't FxA do COPPA/age checking on their end already? So theoretically you can't sign up for a Firefox Account and sign in if you're <13.
Disclaimer: I am not a lawyer. :shockedpikachu:

flozia commented on src/app/(proper_react)/redesign/(authenticated)/user/welcome/EnterInfo.tsx:
```diff
@@ -5,36 +5,79 @@
+  dateOfBirth: ISO8601DateString;
+};
+
+const getAgeFromDateString = (dateOfBirth: ISO8601DateString): number => {
+  const dateNow = new Date();
+  const dateBirth = new Date(dateOfBirth);
+  const dateDelta = new Date(dateNow.valueOf() - dateBirth.valueOf());
+  const unixStartDate = new Date(0);
+
+  return dateDelta.getUTCFullYear() - unixStartDate.getUTCFullYear();
+};
```
Not sure if this is the most bulletproof way to check the age from a date, but I think it is working. This is only for validating the field input and that OneRep will reject a profile that has a birth date under 13. Should we add additional validation to our `/api/v1/user/welcome-scan/create` route @rhelmer @mansaj?

rhelmer replied:
Yes I think we should guard against this on the route as well.

rhelmer replied:
I'd suggest moving this to a shared function so we can use it from the route as well. I think this approach is OK (the new Temporal API or a third-party library like `moment` is easier but I think sticking to standard JS for now is good).

As we already discussed maybe `isEligible()` function could be the function that we call, currently it just checks that the user has not already performed a scan, but checking that they are of the correct age would be useful too.

flozia replied:
Moved the age check into a shared function and am also calling it now additionally in the `/welcome-scan/create` route: [31a6f65](https://github.com/mozilla/blurts-server/pull/3222/commits/31a6f650df5673b5598881022f875d0ea625d83b).

rhelmer commented:
>     2. Make sure you enabled premium in your `.env`:
> 
> 
> ```
> PREMIUM_ENABLED="true"
> ```

Hm, this is only for removal functionality not the scan (users get one free scan), right? Do you want to go ahead and implement+test this now? I think that's OK if so, we don't have the webhook set up yet for this on dev but would be easy to do.

pdehaan commented on src/app/components/server/button.module.scss:
```diff
@@ -19,7 +19,7 @@
 
     &:focus {
-      outline: 3px solid $color-blue-30;
+      outline: $boder-focus-width solid $color-blue-30;
```
globlal: Typo on "border" (looks like 7 instances in this PR).

```suggestion
      outline: $border-focus-width solid $color-blue-30;
```

flozia replied:
Fixed in [d5df068](https://github.com/mozilla/blurts-server/pull/3222/commits/d5df068d91df6ba685c8fef9a895ea4bf53c27d3).

pdehaan commented on src/app/components/server/button.module.scss:
```diff
@@ -31,7 +31,7 @@
 
       &:focus {
-        outline: 3px solid $color-red-30;
+        outline: $boder-focus-width solid $color-red-30;
```
```suggestion
        outline: $border-focus-width solid $color-red-30;
```

pdehaan commented on src/app/tokens.scss:
```diff
@@ -155,6 +155,8 @@ $border-radius-sm: 4px;
 $border-radius-md: 8px;
 $border-radius-lg: 16px;
 
+$boder-focus-width: 3px;
```
```suggestion
$border-focus-width: 3px;
```

pdehaan commented on src/app/components/client/ComboBox.module.scss:
```diff
@@ -20,6 +20,16 @@
+
+      &:focus {
+        border: 1px solid $color-error;
+        outline: $boder-focus-width solid $color-error-focus;
```
```suggestion
        outline: $border-focus-width solid $color-error-focus;
```

pdehaan commented on src/app/components/client/ComboBox.module.scss:
```diff
@@ -20,6 +20,16 @@
+
+    &:focus {
+      border: 1px solid $color-informational-active;
+      outline: $boder-focus-width solid $color-informational-focus;
```
```suggestion
      outline: $border-focus-width solid $color-informational-focus;
```

pdehaan commented on src/app/components/client/InputField.module.scss:
```diff
@@ -19,6 +20,16 @@
+
+      &:focus {
+        border: 1px solid $color-error;
+        outline: $boder-focus-width solid $color-error-focus;
```
```suggestion
        outline: $border-focus-width solid $color-error-focus;
```

pdehaan commented on src/app/components/client/InputField.module.scss:
```diff
@@ -19,6 +20,16 @@
+
+    &:focus {
+      border: 1px solid $color-informational-active;
+      outline: $boder-focus-width solid $color-informational-focus;
```
```suggestion
      outline: $border-focus-width solid $color-informational-focus;
```

flozia commented on src/app/(proper_react)/redesign/(authenticated)/user/welcome/EnterInfo.tsx:
```diff
@@ -63,6 +109,7 @@ export const EnterInfo = (props: Props) => {
       errorMessage: l10n.getString(
         "onboarding-enter-details-input-error-message"
       ),
+      isValid: firstName !== "",
```
That would be good, yes. Added in [d5df068](https://github.com/mozilla/blurts-server/pull/3222/commits/d5df068d91df6ba685c8fef9a895ea4bf53c27d3).

flozia commented on src/app/(proper_react)/redesign/(authenticated)/user/welcome/FindExposures.tsx:
```diff
@@ -0,0 +1,72 @@
+  const progressSteps = 6;
+  const maxProgress = 100;
+  const totalBreachesCount = 672;
+  const scannedBreachesCount = Math.ceil((672 * scanProgress) / 100);
```
Thanks — changed in [d5df068](https://github.com/mozilla/blurts-server/pull/3222/commits/d5df068d91df6ba685c8fef9a895ea4bf53c27d3).

mansaj commented on src/app/api/v1/user/welcome-scan/create/route.ts:
```diff
@@ -0,0 +1,89 @@
+    throw new Error("User has no more manual scans left");
+  }
+
+  const params = await req.json();
```
We can create an interface here for the required params?

rhelmer replied:
They are checked right below, maybe using an interface and type guard for the runtime check would be a better pattern

flozia replied:
Yes! Moved the interface over from `EnterInfo` and sharing it now.

mansaj commented on src/app/api/v1/user/welcome-scan/result/route.ts:
```diff
@@ -0,0 +1,39 @@
+
+import { getLatestOnerepScan } from "../../../../../../db/tables/onerep_scans";
+
+export async function GET(_req: NextRequest) {
```
Since we aren't using `_req`, I think we can just have GET()?

rhelmer replied:
Note that this is all the existing code I landed and not @flozia's fault at all, but yes I agree this is redundant.

flozia replied:
Removed in [31a6f65](https://github.com/mozilla/blurts-server/pull/3222/commits/31a6f650df5673b5598881022f875d0ea625d83b).

flozia commented on src/app/components/client/ProgressBar.module.scss:
```diff
@@ -0,0 +1,38 @@
+      position: absolute;
+      font: $text-title-xs;
+      left: $spacing-lg;
+      top: 52%;
```
Done in [d173203](https://github.com/mozilla/blurts-server/pull/3222/commits/d1732035487e40cd06576680559a6112e99e7f52).

rhelmer reviewed (changes requested):
This is looking great, couple suggestions.

flozia commented:
> > ```
> > PREMIUM_ENABLED="true"
> > ```
> 
> Hm, this is only for removal functionality not the scan (users get one free scan), right? Do you want to go ahead and implement+test this now? I think that's OK if so, we don't have the webhook set up yet for this on dev but would be easy to do.

@rhelmer That’s right — I removed the check again and addressed your other comments as well.

flodolo commented on locales-pending/onboarding.ftl:
```diff
@@ -13,8 +13,8 @@ onboarding-get-started-how-it-works-dialog-heading-line1 = We’re here to prote
-#   $dataBrokerCount (number) - number of scanned data broker sites, e.g. 190
-onboarding-get-started-how-it-works-dialog-step1-content = With just a few key pieces of information, we’ll search for you in all known data breaches and { $dataBrokerCount } major data broker sites. All users get the first scan free.
+#   $dataBrokerTotalCount (number) - number of scanned data broker sites, e.g. 190
+onboarding-get-started-how-it-works-dialog-step1-content = With just a few key pieces of information, we’ll search for you in all known data breaches and { $dataBrokerTotalCount } major data broker sites. All users get the first scan free.
```
Always use the plural form for this type of strings, as it triggers the correct UI in Pontoon.

```
onboarding-get-started-how-it-works-dialog-step1-content = {
    $dataBrokerTotalCount -> 
        *[other] With just a few key pieces of information, we’ll search for you in all known data breaches and { $dataBrokerTotalCount } major data broker sites. All users get the first scan free.
}    
```

flodolo replied:
(technically, you could also decide to provide the correct form for one)

```
onboarding-get-started-how-it-works-dialog-step1-content = {
    $dataBrokerTotalCount -> 
         [one] With just a few key pieces of information, we’ll search for you in all known data breaches and { $dataBrokerTotalCount } major data broker site. All users get the first scan free.
        *[other] With just a few key pieces of information, we’ll search for you in all known data breaches and { $dataBrokerTotalCount } major data broker sites. All users get the first scan free.
}    
```

flodolo replied:
> The count will be always bigger than one.

That is not the point though. We want the plural UI to show up automatically in Pontoon, and a lot of locales have more complex rules than English (and more forms).

flozia replied:
Yes, I understood that this is needed for variables and I added your change request. I mean I just did not add a case for `[one]` or would that be needed always as well?

flodolo replied:
Ah, misunderstood that. `*[other]` is enough 👍🏼

rhelmer commented on .env-dist:
```diff
@@ -75,6 +75,7 @@ HIBP_BREACH_DOMAIN_BLOCKLIST=a-blocked-domain.com,another-blocked-domain.org
 ONEREP_API_BASE=https://api.onerep.com
 ONEREP_API_KEY=
 ONEREP_WEBHOOK_SECRET=
+ONEREP_DATA_BROKER_COUNT=190
```
I think this is fine for now, these don't change that frequently. There is an API for this that we could automatically pull this number from so we don't need to update it manually - https://docs.onerep.com/#operation/getListDatabrokers - I'll file a ticket in the backlog, we should do it as part of background cron jobs IMHO.

rhelmer reviewed (approved):
lgtm, thanks!

flozia commented on src/app/(proper_react)/redesign/(authenticated)/user/welcome/EnterInfo.tsx:
```diff
@@ -5,36 +5,60 @@
 "use client";
```
Yes, I’ll address that.

Happy birthday! 🎂
